File: libraries/lend-shared/src/interfaces/overseer.rs
use fadroma::{
    admin, auth, cosmwasm_std,
    cosmwasm_std::{
        to_binary, Api, Binary, CanonicalAddr, HandleResponse, HumanAddr, InitResponse, Querier,
        QueryRequest, StdError, StdResult, WasmQuery,
    },
    derive_contract::*,
    schemars, Canonize, ContractInstantiationInfo, ContractLink, Decimal256, Humanize, Uint256,
};

use serde::{Deserialize, Serialize};

use crate::core::{AuthMethod, MasterKey, Pagination};
use crate::interfaces::market::Config as MarketConfig;

#[interface(
    component(path = "admin"),
    component(path = "auth", skip(query))
)]
pub trait Overseer {
    #[init]
    fn new(
        admin: Option<HumanAddr>,
        prng_seed: Binary,
        entropy: Binary,
        close_factor: Decimal256,
        // Liquidation incentive
        premium: Decimal256,
        market_contract: ContractInstantiationInfo,
        // Oracle instantiation info
        oracle_contract: ContractInstantiationInfo,
        // Price source for the oracle
        oracle_source: ContractLink<HumanAddr>,
    ) -> StdResult<InitResponse>;

    #[handle]
    fn register_oracle() -> StdResult<HandleResponse>;

    #[handle]
    fn whitelist(config: MarketInitConfig) -> StdResult<HandleResponse>;

    #[handle]
    fn register_market() -> StdResult<HandleResponse>;

    #[handle]
    fn enter(markets: Vec<HumanAddr>) -> StdResult<HandleResponse>;

    #[handle]
    fn exit(market_address: HumanAddr) -> StdResult<HandleResponse>;

    #[handle]
    fn change_market(
        market: HumanAddr,
        ltv_ratio:  Option<Decimal256>,
        symbol: Option<String>
    ) -> StdResult<HandleResponse>;

    #[handle]
    fn change_config(
        premium_rate: Option<Decimal256>,
        close_factor: Option<Decimal256>
    ) -> StdResult<HandleResponse>;

    #[query]
    fn markets(pagination: Pagination) -> StdResult<MarketsResponse>;

    #[query]
    fn market(address: HumanAddr) -> StdResult<Market<HumanAddr>>;

    #[query]
    fn entered_markets(method: OverseerAuth) -> StdResult<Vec<Market<HumanAddr>>>;

    #[query]
    fn oracle_contract() -> StdResult<ContractLink<HumanAddr>>;

    #[query]
    fn account_liquidity(
        method: OverseerAuth,
        market: Option<HumanAddr>,
        block: Option<u64>,
        redeem_amount: Uint256,
        borrow_amount: Uint256,
    ) -> StdResult<AccountLiquidity>;

    #[query]
    fn can_transfer_internal(
        key: MasterKey,
        address: HumanAddr,
        market: HumanAddr,
        block: u64,
        amount: Uint256,
    ) -> StdResult<bool>;

    #[query]
    fn seize_amount(
        borrowed: HumanAddr,
        collateral: HumanAddr,
        repay_amount: Uint256,
    ) -> StdResult<Uint256>;

    #[query]
    fn config() -> StdResult<Config>;
}

pub type OverseerAuth = AuthMethod<OverseerPermissions>;

#[derive(Serialize, Deserialize, Clone, PartialEq, schemars::JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub enum OverseerPermissions {
    AccountInfo,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, schemars::JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub struct AccountLiquidity {
    /// The USD value borrowable by the user, before it reaches liquidation.
    pub liquidity: Uint256,
    /// If > 0 the account is currently below the collateral requirement and is subject to liquidation.
    pub shortfall: Uint256,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, schemars::JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub struct Config {
    /// The discount on collateral that a liquidator receives.
    premium: Decimal256,
    /// The percentage of a liquidatable account's borrow that can be repaid in a single liquidate transaction.
    /// If a user has multiple borrowed assets, the close factor applies to any single borrowed asset,
    /// not the aggregated value of a user’s outstanding borrowing.
    close_factor: Decimal256
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct MarketsResponse {
    pub entries: Vec<Market<HumanAddr>>,
    /// The total number of entries stored by the contract.
    pub total: u64
}

#[derive(Serialize, Deserialize, schemars::JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub struct Market<A> {
    pub contract: ContractLink<A>,
    /// The symbol of the underlying asset.
    pub symbol: String,
    /// The percentage rate at which tokens can be borrowed given the size of the collateral.
    pub ltv_ratio: Decimal256,
}

#[derive(Serialize, Deserialize, schemars::JsonSchema, Debug)]
#[serde(rename_all = "snake_case")]
pub struct MarketInitConfig {
    // Underlying asset address.
    pub underlying_asset: ContractLink<HumanAddr>,
    /// The percentage rate at which tokens can be borrowed given the size of the collateral.
    pub ltv_ratio: Decimal256,
    // Interest model contract address.
    pub interest_model_contract: ContractLink<HumanAddr>,
    pub config: MarketConfig,
    /// Symbol of the underlying asset. Must be the same as what the oracle expects.
    pub token_symbol: String,
    pub prng_seed: Binary,
    pub entropy: Binary
}

impl Config {
    pub fn new(premium: Decimal256, close_factor: Decimal256) -> StdResult<Self> {
        Self::validate_close_factor(&close_factor)?;
        Self::validate_premium(&premium)?;

        Ok(Self {
            premium,
            close_factor
        })
    }

    pub fn set_close_factor(&mut self, new: Decimal256) -> StdResult<()> {
        Self::validate_close_factor(&new)?;

        self.close_factor = new;

        Ok(())
    }

    #[inline]
    pub fn close_factor(&self) -> Decimal256 {
        self.close_factor
    }

    pub fn set_premium(&mut self, new: Decimal256) -> StdResult<()> {
        Self::validate_premium(&new)?;

        self.premium = new;

        Ok(())
    }

    #[inline]
    pub fn premium(&self) -> Decimal256 {
        self.premium
    }

    fn validate_premium(premium: &Decimal256) -> StdResult<()> {
        if *premium < Decimal256::one() {
            return Err(StdError::generic_err("Premium rate cannot be less than 1."))
        } else {
            Ok(())
        }
    }

    fn validate_close_factor(close_factor: &Decimal256) -> StdResult<()> {
        if *close_factor > Decimal256::one() ||
            *close_factor < Decimal256(50000000000000000u128.into()) {
            return Err(StdError::generic_err("Close factor must be between 0.05 and 1"))
        } else {
            Ok(())
        }
    }
}

impl<T> Market<T> {
    pub fn validate(&self) -> StdResult<()> {
        if self.ltv_ratio > Decimal256::one() {
            return Err(StdError::generic_err("LTV ratio must be between 0 and 1."));
        }

        Ok(())
    }
}

impl Canonize for Market<HumanAddr> {
    type Output = Market<CanonicalAddr>;

    fn canonize(self, api: &impl Api) -> StdResult<Self::Output> {
        Ok(Market {
            symbol: self.symbol,
            contract: self.contract.canonize(api)?,
            ltv_ratio: self.ltv_ratio,
        })
    }
}

impl Humanize for Market<CanonicalAddr> {
    type Output = Market<HumanAddr>;

    fn humanize(self, api: &impl Api) -> StdResult<Self::Output> {
        Ok(Market {
            symbol: self.symbol,
            contract: self.contract.humanize(api)?,
            ltv_ratio: self.ltv_ratio,
        })
    }
}

pub fn query_account_liquidity(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
    key: MasterKey,
    address: HumanAddr,
    market: Option<HumanAddr>,
    block: Option<u64>,
    redeem_amount: Uint256,
    borrow_amount: Uint256,
) -> StdResult<AccountLiquidity> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::AccountLiquidity {
            method: OverseerAuth::Internal { key, address },
            market,
            block,
            redeem_amount,
            borrow_amount,
        })?,
    }))
}

pub fn query_can_transfer(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
    key: MasterKey,
    address: HumanAddr,
    market: HumanAddr,
    block: u64,
    amount: Uint256,
) -> StdResult<bool> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::CanTransferInternal {
            key,
            address,
            market,
            block,
            amount,
        })?,
    }))
}

pub fn query_market(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
    address: HumanAddr,
) -> StdResult<Market<HumanAddr>> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::Market { address })?,
    }))
}

pub fn query_config(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
) -> StdResult<Config> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::Config {})?,
    }))
}

pub fn query_seize_amount(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
    borrowed: HumanAddr,
    collateral: HumanAddr,
    repay_amount: Uint256,
) -> StdResult<Uint256> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::SeizeAmount {
            borrowed,
            collateral,
            repay_amount,
        })?,
    }))
}

pub fn query_entered_markets(
    querier: &impl Querier,
    overseer: ContractLink<HumanAddr>,
    key: MasterKey,
    address: HumanAddr
) -> StdResult<Vec<Market<HumanAddr>>> {
    querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: overseer.address,
        callback_code_hash: overseer.code_hash,
        msg: to_binary(&QueryMsg::EnteredMarkets {
            method: AuthMethod::Internal {
                key,
                address
            }
        })?,
    }))
}

File: frontends/siennajs/lib/lend/overseer.ts
import { SmartContract, Querier } from '../contract'
import { Fee, Address, Pagination, ContractInfo, Decimal256, Uint256 } from '../core'
import { ViewingKeyComponentExecutor } from '../executors/viewing_key_executor'
import { LendAuth } from './auth'
import { PaginatedResponse } from '.'

import { ExecuteResult } from 'secretjs'

export interface Market {
    contract: ContractInfo,
    /**
     * The symbol of the underlying asset. Note that this is the same as the symbol
     * that the oracle expects, not what the actual token has in its storage.
     */
    symbol: string,
    /**
     * The percentage rate at which tokens can be borrowed given the size of the collateral.
     */
    ltv_ratio: Decimal256
}

/**
 * One of the fields will always be 0, depending on the state of the account.
 */
export interface AccountLiquidity {
    /**
     * The USD value borrowable by the user, before it reaches liquidation.
     */
    liquidity: Uint256,
    /**
     * If > 0 the account is currently below the collateral requirement and is subject to liquidation.
     */
    shortfall: Uint256
}

export interface OverseerConfig {
    /**
     * The discount on collateral that a liquidator receives.
     */
    premium: Decimal256,
    /**
     * The percentage of a liquidatable account's borrow that can be repaid in a single liquidate transaction.
     * If a user has multiple borrowed assets, the close factor applies to any single borrowed asset,
     * not the aggregated value of a user’s outstanding borrowing.
     */
    close_factor: Decimal256
}

export type OverseerPermissions = 'account_info'

export class OverseerContract extends SmartContract<OverseerExecutor, OverseerQuerier> {
    exec(fee?: Fee, memo?: string): OverseerExecutor {
        return new OverseerExecutor(
            this.address,
            this.execute_client,
            fee,
            memo
        )
    }

    query(): OverseerQuerier {
        return new OverseerQuerier(this.address, this.query_client)
    }
}

class OverseerExecutor extends ViewingKeyComponentExecutor {
    async enter_markets(markets: Address[]): Promise<ExecuteResult> {
        const msg = {
            enter: {
                markets
            }
        }

        return this.run(msg, '40000')
    }

    async exit_markets(market_address: Address): Promise<ExecuteResult> {
        const msg = {
            exit: {
                market_address
            }
        }

        return this.run(msg, '50000')
    }
}

class OverseerQuerier extends Querier {
    /**
     * Max limit per page is `30`.
     */
    async markets(pagination: Pagination): Promise<PaginatedResponse<Market>> {
        const msg = {
            markets: {
                pagination
            }
        }

        return this.run(msg)
    }

    async market(address: Address): Promise<Market> {
        const msg = {
            market: {
                address
            }
        }

        return this.run(msg)
    }

    async entered_markets(auth: LendAuth): Promise<Market[]> {
        const msg = {
            entered_markets: {
                method: await auth.create_method<OverseerPermissions>(this.address, 'account_info')
            }
        }

        return this.run(msg)
    }

    async current_liquidity(auth: LendAuth): Promise<AccountLiquidity> {
        const msg = {
            account_liquidity: {
                block: (await this.client.getBlock()).header.height,
                method: await auth.create_method<OverseerPermissions>(this.address, 'account_info'),
                market: null,
                redeem_amount: '0',
                borrow_amount: '0'
            }
        }

        return this.run(msg)
    }

    /**
     * The hypothetical liquidity after a redeem operation from a market.
     * 
     * @param method 
     * @param market - The market to redeem from. Must have been entered that market.
     * @param redeem_amount - The amount to redeem.
     */
    async liquidity_after_redeem(
        auth: LendAuth,
        market: Address,
        redeem_amount: Uint256
    ): Promise<AccountLiquidity> {
        const msg = {
            account_liquidity: {
                block: (await this.client.getBlock()).header.height,
                method: await auth.create_method<OverseerPermissions>(this.address, 'account_info'),
                market,
                redeem_amount,
                borrow_amount: '0'
            }
        }

        return this.run(msg)
    }

    /**
     * The hypothetical liquidity after a borrow operation from a market.
     * 
     * @param method 
     * @param market - The market to borrow from. Must have been entered that market.
     * @param borrow_amount - The amount to borrow.
     */
    async liquidity_after_borrow(
        auth: LendAuth,
        market: Address,
        borrow_amount: Uint256
    ): Promise<AccountLiquidity> {
        const msg = {
            account_liquidity: {
                block: (await this.client.getBlock()).header.height,
                method: await auth.create_method<OverseerPermissions>(this.address, 'account_info'),
                market,
                redeem_amount: '0',
                borrow_amount
            }
        }

        return this.run(msg)
    }

    /**
     * The hypothetical amount that will be seized from a liquidation.
     * 
     * @param borrowed - The market that is being liquidated.
     * @param collateral - The slToken collateral to be seized. 
     * @param repay_amount - The liquidation amount.
     */
    async seize_amount(
        borrowed: Address,
        collateral: Address,
        repay_amount: Uint256
    ): Promise<Uint256> {
        const msg = {
            seize_amount: {
                borrowed,
                collateral,
                repay_amount
            }
        }

        return this.run(msg)
    }

    async config(): Promise<OverseerConfig> {
        const msg = {
            config: { }
        }

        return this.run(msg)
    }
}
